# é–¢æ•°å‹ãƒ—ãƒ­ã‚°ãƒ©ãƒŸãƒ³ã‚° (FP) ãƒ«ãƒ¼ãƒ«

ã“ã®ãƒ—ãƒ­ã‚¸ã‚§ã‚¯ãƒˆã§ã¯ã€**å®Ÿç”¨çš„ãªé–¢æ•°å‹ãƒ—ãƒ­ã‚°ãƒ©ãƒŸãƒ³ã‚°**ã®åŸå‰‡ã‚’æ¡ç”¨ã—ã¾ã™ã€‚
ç´”ç²‹ãªFPã§ã¯ãªãã€TypeScriptã¨Reactã®ã‚¨ã‚³ã‚·ã‚¹ãƒ†ãƒ ã«é©ã—ãŸã€ãƒãƒ©ãƒ³ã‚¹ã®å–ã‚ŒãŸã‚¢ãƒ—ãƒ­ãƒ¼ãƒã‚’ç›®æŒ‡ã—ã¾ã™ã€‚

## æ ¸å¿ƒåŸå‰‡

### 1. ã‚¤ãƒŸãƒ¥ãƒ¼ã‚¿ãƒ“ãƒªãƒ†ã‚£ï¼ˆä¸å¤‰æ€§ï¼‰

**åŸå‰‡**: ãƒ‡ãƒ¼ã‚¿ã‚’å¤‰æ›´ã›ãšã€æ–°ã—ã„ãƒ‡ãƒ¼ã‚¿ã‚’ä½œæˆã™ã‚‹

#### âœ… Good - ãƒ—ãƒ­ã‚¸ã‚§ã‚¯ãƒˆå®Ÿä¾‹
```typescript
// TechStackService.ts - ãƒ•ã‚£ãƒ«ã‚¿ãƒªãƒ³ã‚°
static filterByCategories(categories: TechCategory[]): TechItem[] {
  return TECH_STACK_LIST.filter((tech) => categories.includes(tech.category));
}

// useTechSelection.ts - Stateæ›´æ–°
const toggle = useCallback((tech: TechItem) => {
  setSelected((prev) =>
    prev.some((item) => item.id === tech.id)
      ? prev.filter((item) => item.id !== tech.id)  // å‰Šé™¤
      : [...prev, tech]                              // è¿½åŠ 
  );
}, []);

// ãƒã‚¹ãƒˆã—ãŸã‚ªãƒ–ã‚¸ã‚§ã‚¯ãƒˆæ›´æ–°
const updateTheme = (theme: Theme, mode: 'light' | 'dark') => ({
  ...theme,
  colors: {
    ...theme.colors,
    mode,
  },
});
```

#### âŒ Bad - ãƒŸãƒ¥ãƒ¼ãƒ†ãƒ¼ã‚·ãƒ§ãƒ³
```typescript
// State ã‚’ç›´æ¥å¤‰æ›´
const toggle = (tech: TechItem) => {
  selected.push(tech); // âŒ ç›´æ¥å¤‰æ›´
  setSelected(selected); // âŒ åŒã˜å‚ç…§
};

// é…åˆ—ãƒ¡ã‚½ãƒƒãƒ‰ã®èª¤ç”¨
const items = [...TECH_STACK_LIST];
items.sort(); // âŒ å…ƒã®é…åˆ—ã‚’å¤‰æ›´
return items;

// ã‚ªãƒ–ã‚¸ã‚§ã‚¯ãƒˆã®ç›´æ¥å¤‰æ›´
tech.name = 'Updated'; // âŒ
```

#### ğŸ’¡ ãƒ‘ãƒ•ã‚©ãƒ¼ãƒãƒ³ã‚¹ã®ãƒ’ãƒ³ãƒˆ
```typescript
// å¤§é‡ãƒ‡ãƒ¼ã‚¿ã®å ´åˆã€é©åˆ‡ãªæœ€é©åŒ–ã‚’
// âŒ æ¯å›å…¨é…åˆ—ã‚’ã‚³ãƒ”ãƒ¼ï¼ˆå¤§é‡ãƒ‡ãƒ¼ã‚¿ã§éåŠ¹ç‡ï¼‰
const addMany = (items: TechItem[], newItems: TechItem[]) => {
  let result = items;
  newItems.forEach(item => {
    result = [...result, item]; // O(nÂ²)
  });
  return result;
};

// âœ… ä¸€åº¦ã«ã¾ã¨ã‚ã¦ã‚³ãƒ”ãƒ¼
const addMany = (items: TechItem[], newItems: TechItem[]) =>
  [...items, ...newItems]; // O(n)
```

---

### 2. ç´”ç²‹é–¢æ•°ï¼ˆPure Functionsï¼‰

**åŸå‰‡**: åŒã˜å…¥åŠ› â†’ åŒã˜å‡ºåŠ›ã€å‰¯ä½œç”¨ãªã—

#### âœ… Good - ã‚µãƒ¼ãƒ“ã‚¹å±¤ã®ç´”ç²‹é–¢æ•°
```typescript
// TechStackService.ts
static getById(id: TechId): TechItem | undefined {
  return TECH_STACK[id as keyof typeof TECH_STACK];
}

static search(keyword: string): TechItem[] {
  const query = keyword.toLowerCase();
  return TECH_STACK_LIST.filter((tech) =>
    tech.name.toLowerCase().includes(query)
  );
}

// MarkdownService.ts - ç´”ç²‹ãªå¤‰æ›ãƒ­ã‚¸ãƒƒã‚¯
static generate(
  selectedTechs: TechItem[],
  format: OutputFormat,
  iconSize: number
): string {
  const icons = selectedTechs
    .map((tech) => this.generateIcon(tech, format, iconSize))
    .join(' ');

  return `<p align="center">\n  ${icons}\n</p>`;
}
```

#### âœ… å‰¯ä½œç”¨ã®åˆ†é›¢
```typescript
// âŒ ç´”ç²‹é–¢æ•°å†…ã§å‰¯ä½œç”¨
function saveAndReturn(tech: TechItem): TechItem {
  localStorage.setItem('tech', JSON.stringify(tech)); // âŒ å‰¯ä½œç”¨
  console.log('Saved:', tech); // âŒ å‰¯ä½œç”¨
  return tech;
}

// âœ… å‰¯ä½œç”¨ã‚’åˆ†é›¢
function saveTech(tech: TechItem): void {
  localStorage.setItem('tech', JSON.stringify(tech));
  console.log('Saved:', tech);
}

function processTech(tech: TechItem): TechItem {
  // ç´”ç²‹ãªå‡¦ç†ã®ã¿
  return { ...tech, processedAt: Date.now() };
}

// ä½¿ç”¨æ™‚
const processed = processTech(tech);
saveTech(processed); // å‰¯ä½œç”¨ã¯æ˜ç¤ºçš„ã«å®Ÿè¡Œ
```

#### ğŸ’¡ è¨±å®¹ã•ã‚Œã‚‹å‰¯ä½œç”¨
```typescript
// âœ… React Hookså†…ã®å‰¯ä½œç”¨ã¯é©åˆ‡
useEffect(() => {
  // å‰¯ä½œç”¨ã¯useEffectã«é›†ç´„
  localStorage.setItem('selected', JSON.stringify(selected));
}, [selected]);

// âœ… ã‚¤ãƒ™ãƒ³ãƒˆãƒãƒ³ãƒ‰ãƒ©å†…ã®å‰¯ä½œç”¨ã‚‚é©åˆ‡
const handleCopy = async () => {
  await navigator.clipboard.writeText(markdown);
  toast({ title: "Copied!" });
};
```

---

### 3. å‹ã«ã‚ˆã‚‹è¡¨ç¾åŠ›ã®å‘ä¸Š

**TypeScriptã®å‹ã‚·ã‚¹ãƒ†ãƒ ã‚’æ´»ç”¨ã—ã€å®Ÿè¡Œæ™‚ã‚¨ãƒ©ãƒ¼ã‚’é˜²ã**

#### âœ… Good - å‹ã§çŠ¶æ…‹ã‚’è¡¨ç¾
```typescript
// Eitherå‹ - ã‚¨ãƒ©ãƒ¼ãƒãƒ³ãƒ‰ãƒªãƒ³ã‚°ã‚’å‹ã§è¡¨ç¾
type Either<E, A> =
  | { _tag: 'Left'; left: E }
  | { _tag: 'Right'; right: A };

const left = <E, A>(e: E): Either<E, A> =>
  ({ _tag: 'Left', left: e });
const right = <E, A>(a: A): Either<E, A> =>
  ({ _tag: 'Right', right: a });

// ä½¿ç”¨ä¾‹
async function validateIcon(id: IconifyId): Promise<Either<string, boolean>> {
  try {
    const response = await fetch(`https://api.iconify.design/${id}.svg`);
    return right(response.ok);
  } catch (error) {
    return left(error instanceof Error ? error.message : 'Unknown error');
  }
}

// ãƒ‘ã‚¿ãƒ¼ãƒ³ãƒãƒƒãƒãƒ³ã‚°é¢¨ã«å‡¦ç†
const result = await validateIcon('logos:react');
if (result._tag === 'Left') {
  console.error('Validation failed:', result.left);
} else {
  console.log('Valid:', result.right);
}
```

#### âœ… Optionå‹ - null/undefinedã‚’å‹å®‰å…¨ã«
```typescript
type Option<T> = T | undefined;

// ãƒ˜ãƒ«ãƒ‘ãƒ¼é–¢æ•°
const map = <T, U>(opt: Option<T>, fn: (value: T) => U): Option<U> =>
  opt === undefined ? undefined : fn(opt);

const getOrElse = <T>(opt: Option<T>, defaultValue: T): T =>
  opt === undefined ? defaultValue : opt;

// ä½¿ç”¨ä¾‹
const tech = TechStackService.getById(id); // Option<TechItem>
const name = map(tech, t => t.name);
const displayName = getOrElse(name, 'Unknown');
```

#### âœ… åˆ¤åˆ¥å…±ç”¨ä½“ï¼ˆDiscriminated Unionï¼‰
```typescript
// çŠ¶æ…‹ã‚’å‹ã§æ˜ç¢ºã«è¡¨ç¾
type LoadingState<T> =
  | { status: 'idle' }
  | { status: 'loading' }
  | { status: 'success'; data: T }
  | { status: 'error'; error: string };

// å‹å®‰å…¨ãªçŠ¶æ…‹ç®¡ç†
function render<T>(state: LoadingState<T>): React.ReactNode {
  switch (state.status) {
    case 'idle':
      return <div>Click to load</div>;
    case 'loading':
      return <div>Loading...</div>;
    case 'success':
      return <div>{state.data}</div>; // data ã¯å‹å®‰å…¨
    case 'error':
      return <div>Error: {state.error}</div>; // error ã¯å‹å®‰å…¨
  }
}
```

---

### 4. é–¢æ•°åˆæˆã¨ãƒ‘ã‚¤ãƒ—ãƒ©ã‚¤ãƒ³

**å°ã•ãªé–¢æ•°ã‚’çµ„ã¿åˆã‚ã›ã¦è¤‡é›‘ãªå‡¦ç†ã‚’æ§‹ç¯‰**

#### âœ… Good - å®Ÿè·µçš„ãªãƒ‘ã‚¤ãƒ—ãƒ©ã‚¤ãƒ³
```typescript
// ãƒ¦ãƒ¼ãƒ†ã‚£ãƒªãƒ†ã‚£é–¢æ•°
const pipe = <T>(...fns: Array<(arg: T) => T>) =>
  (value: T) => fns.reduce((acc, fn) => fn(acc), value);

// å°ã•ãªç´”ç²‹é–¢æ•°ã‚’å®šç¾©
const filterByCategory = (category: TechCategory) =>
  (items: TechItem[]) => items.filter(item => item.category === category);

const sortByName = (items: TechItem[]) =>
  [...items].sort((a, b) => a.name.localeCompare(b.name));

const take = (n: number) => <T>(items: T[]) => items.slice(0, n);

const mapToNames = (items: TechItem[]) => items.map(item => item.name);

// åˆæˆã—ã¦è¤‡é›‘ãªå‡¦ç†ã‚’æ§‹ç¯‰
const getTopFrameworkNames = pipe(
  filterByCategory(TechCategory.Framework),
  sortByName,
  take(10),
  mapToNames
);

const result = getTopFrameworkNames(TECH_STACK_LIST);
```

#### âœ… å®Ÿéš›ã®ãƒ—ãƒ­ã‚¸ã‚§ã‚¯ãƒˆã§ã®æ´»ç”¨
```typescript
// entities/tech/lib/utils.ts ã«é…ç½®
export const createTechFilter = (predicate: (tech: TechItem) => boolean) =>
  (items: TechItem[]) => items.filter(predicate);

export const sortTech = (compareFn: (a: TechItem, b: TechItem) => number) =>
  (items: TechItem[]) => [...items].sort(compareFn);

// features/tech-selection ã§ä½¿ç”¨
const filterActive = createTechFilter(tech => tech.isActive);
const sortByName = sortTech((a, b) => a.name.localeCompare(b.name));

const getActiveTechsSorted = pipe(
  filterActive,
  sortByName
);
```

---

### 5. é«˜éšé–¢æ•°ã®æ´»ç”¨

#### âœ… ã‚«ãƒªãƒ¼åŒ–ï¼ˆCurryingï¼‰
```typescript
// ã‚«ãƒªãƒ¼åŒ–ã•ã‚ŒãŸãƒ•ã‚£ãƒ«ã‚¿ãƒ¼
const filterBy = <T>(key: keyof T) =>
  (value: T[typeof key]) =>
    (items: T[]) => items.filter(item => item[key] === value);

// éƒ¨åˆ†é©ç”¨ã§å†åˆ©ç”¨å¯èƒ½ãªé–¢æ•°ã‚’ä½œæˆ
const filterByCategory = filterBy<TechItem>('category');
const getFrameworks = filterByCategory(TechCategory.Framework);
const getLibraries = filterByCategory(TechCategory.Library);

// ä½¿ç”¨
const frameworks = getFrameworks(TECH_STACK_LIST);
const libraries = getLibraries(TECH_STACK_LIST);
```

#### âœ… é–¢æ•°ãƒ•ã‚¡ã‚¯ãƒˆãƒªãƒ¼
```typescript
// Markdownç”Ÿæˆã®ãƒ•ã‚¡ã‚¯ãƒˆãƒªãƒ¼
const createMarkdownGenerator = (format: OutputFormat) =>
  (iconSize: number) =>
    (techs: TechItem[]) =>
      MarkdownService.generate(techs, format, iconSize);

// ç‰¹å®šã®è¨­å®šã§å›ºå®š
const generateSingle = createMarkdownGenerator('single');
const generateSingle48 = generateSingle(48);

// ä½¿ç”¨
const markdown = generateSingle48(selectedTechs);
```

---

## React ã§ã® FP å®Ÿè·µ

### 1. Hooks ã¯ãƒ­ã‚¸ãƒƒã‚¯ã®ã‚«ãƒ—ã‚»ãƒ«åŒ–

#### âœ… Good - ç´”ç²‹ãªãƒ­ã‚¸ãƒƒã‚¯ã‚’Hooksã«
```typescript
// features/tech-selection/model/useTechSelection.ts
export function useTechSelection(initialItems: TechItem[] = []) {
  const [selected, setSelected] = useState<TechItem[]>(initialItems);

  // ç´”ç²‹ãªæ›´æ–°ãƒ­ã‚¸ãƒƒã‚¯
  const toggle = useCallback((tech: TechItem) => {
    setSelected((prev) =>
      prev.some((item) => item.id === tech.id)
        ? prev.filter((item) => item.id !== tech.id)
        : [...prev, tech]
    );
  }, []);

  const selectAll = useCallback((items: TechItem[]) => {
    setSelected(items);
  }, []);

  const clear = useCallback(() => {
    setSelected([]);
  }, []);

  // æ´¾ç”ŸçŠ¶æ…‹ã¯ useMemo ã§è¨ˆç®—
  const stats = useMemo(
    () => ({
      count: selected.length,
      isEmpty: selected.length === 0,
      ids: selected.map((tech) => tech.id),
    }),
    [selected]
  );

  return { selected, toggle, selectAll, clear, stats };
}
```

### 2. ã‚³ãƒ³ãƒãƒ¼ãƒãƒ³ãƒˆã¯ç´”ç²‹é–¢æ•°ã®ã‚ˆã†ã«

#### âœ… Good - ãƒ—ãƒ¬ã‚¼ãƒ³ãƒ†ãƒ¼ã‚·ãƒ§ãƒŠãƒ«ã‚³ãƒ³ãƒãƒ¼ãƒãƒ³ãƒˆ
```typescript
// ç´”ç²‹ãªã‚³ãƒ³ãƒãƒ¼ãƒãƒ³ãƒˆ - props â†’ UI
interface TechCardProps {
  tech: TechItem;
  isSelected: boolean;
  onToggle: (tech: TechItem) => void;
}

export function TechCard({ tech, isSelected, onToggle }: TechCardProps) {
  return (
    <div onClick={() => onToggle(tech)}>
      <img src={IconifyService.getIconUrl(tech.iconify)} alt={tech.name} />
      <span>{tech.name}</span>
      {isSelected && <CheckIcon />}
    </div>
  );
}
```

### 3. å‰¯ä½œç”¨ã®æ˜ç¤ºçš„ãªåˆ†é›¢

#### âœ… Good - useEffect ã§å‰¯ä½œç”¨ã‚’é›†ç´„
```typescript
function TechSelector() {
  const { selected, toggle } = useTechSelection();
  const [markdown, setMarkdown] = useState('');

  // å‰¯ä½œç”¨1: LocalStorage ã¸ã®ä¿å­˜
  useEffect(() => {
    localStorage.setItem('selected', JSON.stringify(selected));
  }, [selected]);

  // å‰¯ä½œç”¨2: Analytics
  useEffect(() => {
    if (selected.length > 0) {
      analytics.track('tech_selected', { count: selected.length });
    }
  }, [selected.length]);

  // å‰¯ä½œç”¨3: Markdownç”Ÿæˆï¼ˆéåŒæœŸï¼‰
  useEffect(() => {
    const md = MarkdownService.generate(selected, 'single', 48);
    setMarkdown(md);
  }, [selected]);

  return (
    // ç´”ç²‹ãªãƒ¬ãƒ³ãƒ€ãƒªãƒ³ã‚°
    <div>{/* ... */}</div>
  );
}
```

---

## å®Ÿè·µçš„ãƒ‘ã‚¿ãƒ¼ãƒ³é›†

### Pattern 1: Maybe/Option ãƒ‘ã‚¿ãƒ¼ãƒ³
```typescript
// shared/lib/maybe.ts
export const mapOption = <T, U>(
  value: T | undefined,
  fn: (v: T) => U
): U | undefined => (value === undefined ? undefined : fn(value));

export const flatMapOption = <T, U>(
  value: T | undefined,
  fn: (v: T) => U | undefined
): U | undefined => (value === undefined ? undefined : fn(value));

// ä½¿ç”¨ä¾‹
const tech = TechStackService.getById(id);
const iconUrl = mapOption(tech, t => IconifyService.getIconUrl(t.iconify));
const displayUrl = iconUrl ?? '/default-icon.svg';
```

### Pattern 2: Result/Either ãƒ‘ã‚¿ãƒ¼ãƒ³
```typescript
// shared/lib/result.ts
export type Result<T, E = Error> =
  | { ok: true; value: T }
  | { ok: false; error: E };

export const ok = <T>(value: T): Result<T> => ({ ok: true, value });
export const err = <E>(error: E): Result<never, E> => ({ ok: false, error });

export const mapResult = <T, U, E>(
  result: Result<T, E>,
  fn: (value: T) => U
): Result<U, E> =>
  result.ok ? ok(fn(result.value)) : result;

// ä½¿ç”¨ä¾‹
async function fetchTechData(id: string): Promise<Result<TechItem, string>> {
  try {
    const response = await fetch(`/api/tech/${id}`);
    if (!response.ok) {
      return err(`HTTP error: ${response.status}`);
    }
    const data = await response.json();
    return ok(data);
  } catch (error) {
    return err(error instanceof Error ? error.message : 'Unknown error');
  }
}
```

### Pattern 3: é…åˆ—ã®ä¾¿åˆ©ãªãƒ˜ãƒ«ãƒ‘ãƒ¼
```typescript
// shared/lib/array.ts
export const groupBy = <T>(
  items: T[],
  key: keyof T
): Record<string, T[]> =>
  items.reduce((acc, item) => {
    const group = String(item[key]);
    return {
      ...acc,
      [group]: [...(acc[group] || []), item],
    };
  }, {} as Record<string, T[]>);

export const uniqueBy = <T>(items: T[], key: keyof T): T[] => {
  const seen = new Set();
  return items.filter(item => {
    const k = item[key];
    if (seen.has(k)) return false;
    seen.add(k);
    return true;
  });
};

// ä½¿ç”¨ä¾‹
const grouped = groupBy(TECH_STACK_LIST, 'category');
const unique = uniqueBy(items, 'id');
```

---

## é¿ã‘ã‚‹ã¹ãã“ã¨

### âŒ éåº¦ãªæŠ½è±¡åŒ–
```typescript
// âŒ ã‚„ã‚Šã™ã - èª­ã¿ã¥ã‚‰ã„
const compose = <A, B, C>(f: (b: B) => C, g: (a: A) => B) =>
  (a: A) => f(g(a));

const map = <T, U>(fn: (t: T) => U) =>
  (arr: T[]) => arr.map(fn);

const filter = <T>(pred: (t: T) => boolean) =>
  (arr: T[]) => arr.filter(pred);

// âœ… ã‚·ãƒ³ãƒ—ãƒ«ã«
const getActiveFrameworks = (items: TechItem[]) =>
  items
    .filter(item => item.category === TechCategory.Framework)
    .filter(item => item.isActive);
```

### âŒ ä¸å¿…è¦ãªã‚«ãƒªãƒ¼åŒ–
```typescript
// âŒ 1å›ã—ã‹ä½¿ã‚ãªã„ã®ã«ã‚«ãƒªãƒ¼åŒ–
const add = (a: number) => (b: number) => a + b;
const result = add(1)(2);

// âœ… æ™®é€šã®é–¢æ•°ã§ååˆ†
const add = (a: number, b: number) => a + b;
const result = add(1, 2);
```

### âœ… ãƒãƒ©ãƒ³ã‚¹ã®å–ã‚ŒãŸã‚¢ãƒ—ãƒ­ãƒ¼ãƒ
```typescript
// å†åˆ©ç”¨æ€§ãŒé«˜ã„ â†’ ã‚«ãƒªãƒ¼åŒ–ãŒæœ‰åŠ¹
const filterBy = <T>(key: keyof T) => (value: T[typeof key]) =>
  (items: T[]) => items.filter(item => item[key] === value);

// ä¸€åº¦ã—ã‹ä½¿ã‚ãªã„ â†’ æ™®é€šã®é–¢æ•°
const getSelectedIds = (items: TechItem[]) => items.map(item => item.id);
```

---

## ãƒ‘ãƒ•ã‚©ãƒ¼ãƒãƒ³ã‚¹ã¨ã®ãƒãƒ©ãƒ³ã‚¹

### é©åˆ‡ãªãƒ¡ãƒ¢åŒ–
```typescript
// âœ… é‡ã„è¨ˆç®—ã¯ãƒ¡ãƒ¢åŒ–
const expensiveComputation = useMemo(() => {
  return items
    .filter(complexFilter)
    .map(expensiveTransform)
    .sort(complexSort);
}, [items]);

// âŒ è»½ã„è¨ˆç®—ã¯ä¸è¦
const simpleValue = useMemo(() => item.name, [item]); // ä¸è¦
```

### ã‚¤ãƒŸãƒ¥ãƒ¼ã‚¿ãƒ–ãƒ«æ“ä½œã®æœ€é©åŒ–
```typescript
// âŒ éåŠ¹ç‡ - O(nÂ²)
let result = [];
for (const item of items) {
  result = [...result, item];
}

// âœ… åŠ¹ç‡çš„ - O(n)
const result = items.slice(); // ã¾ãŸã¯ [...items]

// âœ… å¤§é‡ãƒ‡ãƒ¼ã‚¿ã¯é©åˆ‡ãªæ§‹é€ ã‚’
import { List } from 'immutable';
const list = List(items);
const updated = list.push(newItem); // O(1) ã«è¿‘ã„
```

---

## ã¾ã¨ã‚

### FP ã®åˆ©ç‚¹ï¼ˆã“ã®ãƒ—ãƒ­ã‚¸ã‚§ã‚¯ãƒˆã§ï¼‰
1. **äºˆæ¸¬å¯èƒ½æ€§**: ç´”ç²‹é–¢æ•°ã¯åŒã˜å…¥åŠ›ã§åŒã˜å‡ºåŠ›
2. **ãƒ†ã‚¹ã‚¿ãƒ“ãƒªãƒ†ã‚£**: å‰¯ä½œç”¨ãŒãªã„ãŸã‚ãƒ¦ãƒ‹ãƒƒãƒˆãƒ†ã‚¹ãƒˆå®¹æ˜“
3. **ä¸¦è¡Œå‡¦ç†å®‰å…¨**: ã‚¤ãƒŸãƒ¥ãƒ¼ã‚¿ãƒ–ãƒ«ãªãƒ‡ãƒ¼ã‚¿ã¯ç«¶åˆã—ãªã„
4. **ãƒã‚°ã®å°‘ãªã•**: ãƒŸãƒ¥ãƒ¼ãƒ†ãƒ¼ã‚·ãƒ§ãƒ³ãƒã‚°ã‚’é˜²ã
5. **ãƒªãƒ•ã‚¡ã‚¯ã‚¿ãƒªãƒ³ã‚°å®¹æ˜“**: ä¾å­˜é–¢ä¿‚ãŒæ˜ç¢º

### å®Ÿè·µã‚¬ã‚¤ãƒ‰ãƒ©ã‚¤ãƒ³
âœ… **å¸¸ã«å®Ÿè·µ**
- Stateæ›´æ–°ã¯ã‚¤ãƒŸãƒ¥ãƒ¼ã‚¿ãƒ–ãƒ«ã«
- ç´”ç²‹é–¢æ•°ã‚’å„ªå…ˆ
- map/filter/reduce ã‚’æ´»ç”¨
- å‰¯ä½œç”¨ã¯æ˜ç¤ºçš„ã«åˆ†é›¢

âœ… **çŠ¶æ³ã«å¿œã˜ã¦**
- é–¢æ•°åˆæˆï¼ˆè¤‡é›‘ãªå‡¦ç†ã®å ´åˆï¼‰
- ã‚«ãƒªãƒ¼åŒ–ï¼ˆå†åˆ©ç”¨æ€§ãŒé«˜ã„å ´åˆï¼‰
- Option/Resultå‹ï¼ˆã‚¨ãƒ©ãƒ¼ãƒãƒ³ãƒ‰ãƒªãƒ³ã‚°ï¼‰

âŒ **é¿ã‘ã‚‹**
- éåº¦ãªæŠ½è±¡åŒ–
- ä¸è¦ãªãƒ¡ãƒ¢åŒ–
- èª­ã¿ã¥ã‚‰ã„ã‚³ãƒ¼ãƒ‰
- ãƒ‘ãƒ•ã‚©ãƒ¼ãƒãƒ³ã‚¹ç„¡è¦–

### ãƒ—ãƒ©ã‚°ãƒãƒ†ã‚£ãƒƒã‚¯FP
å®Œç’§ãªFPã‚ˆã‚Šã‚‚ã€**ãƒãƒ¼ãƒ ãŒç†è§£ã—ã‚„ã™ãã€ä¿å®ˆã—ã‚„ã™ã„ã‚³ãƒ¼ãƒ‰**ã‚’ç›®æŒ‡ã—ã¾ã™ã€‚

## å‚è€ƒè³‡æ–™

- [fp-ts](https://gcanti.github.io/fp-ts/) - TypeScript FP ãƒ©ã‚¤ãƒ–ãƒ©ãƒª
- [Ramda](https://ramdajs.com/) - FP ãƒ¦ãƒ¼ãƒ†ã‚£ãƒªãƒ†ã‚£
- [React Hooks FP](https://kentcdodds.com/blog/react-hooks-pitfalls)
- [Immutability in React](https://react.dev/learn/updating-objects-in-state)
